{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7e6bffd2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[90mtreeio v1.26.0 For help: https://yulab-smu.top/treedata-book/\n",
      "\n",
      "If you use the ggtree package suite in published research, please cite\n",
      "the appropriate paper(s):\n",
      "\n",
      "LG Wang, TTY Lam, S Xu, Z Dai, L Zhou, T Feng, P Guo, CW Dunn, BR\n",
      "Jones, T Bradley, H Zhu, Y Guan, Y Jiang, G Yu. treeio: an R package\n",
      "for phylogenetic tree input and output with richly annotated and\n",
      "associated data. Molecular Biology and Evolution. 2020, 37(2):599-603.\n",
      "doi: 10.1093/molbev/msz240\n",
      "\n",
      "Guangchuang Yu, Tommy Tsan-Yuk Lam, Huachen Zhu, Yi Guan. Two methods\n",
      "for mapping and visualizing associated data on phylogeny using ggtree.\n",
      "Molecular Biology and Evolution. 2018, 35(12):3041-3043.\n",
      "doi:10.1093/molbev/msy194\n",
      "\n",
      "G Yu. Data Integration, Manipulation and Visualization of Phylogenetic\n",
      "Trees (1st ed.). Chapman and Hall/CRC. 2022. ISBN: 9781032233574 \u001b[39m\n",
      "\n",
      "\u001b[90mggtree v3.10.1 For help: https://yulab-smu.top/treedata-book/\n",
      "\n",
      "If you use the ggtree package suite in published research, please cite\n",
      "the appropriate paper(s):\n",
      "\n",
      "Guangchuang Yu, David Smith, Huachen Zhu, Yi Guan, Tommy Tsan-Yuk Lam.\n",
      "ggtree: an R package for visualization and annotation of phylogenetic\n",
      "trees with their covariates and other associated data. Methods in\n",
      "Ecology and Evolution. 2017, 8(1):28-36. doi:10.1111/2041-210X.12628\n",
      "\n",
      "LG Wang, TTY Lam, S Xu, Z Dai, L Zhou, T Feng, P Guo, CW Dunn, BR\n",
      "Jones, T Bradley, H Zhu, Y Guan, Y Jiang, G Yu. treeio: an R package\n",
      "for phylogenetic tree input and output with richly annotated and\n",
      "associated data. Molecular Biology and Evolution. 2020, 37(2):599-603.\n",
      "doi: 10.1093/molbev/msz240\n",
      "\n",
      "G Yu. Data Integration, Manipulation and Visualization of Phylogenetic\n",
      "Trees (1st ed.). Chapman and Hall/CRC. 2022. ISBN: 9781032233574 \u001b[39m\n",
      "\n",
      "\n",
      "Attaching package: ‘ape’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:ggtree’:\n",
      "\n",
      "    rotate\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:treeio’:\n",
      "\n",
      "    drop.tip\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "========================================\n",
      "ComplexHeatmap version 2.18.0\n",
      "Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/\n",
      "Github page: https://github.com/jokergoo/ComplexHeatmap\n",
      "Documentation: http://jokergoo.github.io/ComplexHeatmap-reference\n",
      "\n",
      "If you use it in published research, please cite either one:\n",
      "- Gu, Z. Complex Heatmap Visualization. iMeta 2022.\n",
      "- Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional \n",
      "    genomic data. Bioinformatics 2016.\n",
      "\n",
      "\n",
      "The new InteractiveComplexHeatmap package can directly export static \n",
      "complex heatmaps into an interactive Shiny app with zero effort. Have a try!\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(ComplexHeatmap))\n",
      "========================================\n",
      "\n",
      "\n",
      "========================================\n",
      "circlize version 0.4.16\n",
      "CRAN page: https://cran.r-project.org/package=circlize\n",
      "Github page: https://github.com/jokergoo/circlize\n",
      "Documentation: https://jokergoo.github.io/circlize_book/book/\n",
      "\n",
      "If you use it in published research, please cite:\n",
      "Gu, Z. circlize implements and enhances circular visualization\n",
      "  in R. Bioinformatics 2014.\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(circlize))\n",
      "========================================\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘circlize’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:ape’:\n",
      "\n",
      "    degree\n",
      "\n",
      "\n",
      "\u001b[90mggtreeExtra v1.12.0 For help: https://yulab-smu.top/treedata-book/\n",
      "\n",
      "If you use the ggtree package suite in published research, please cite\n",
      "the appropriate paper(s):\n",
      "\n",
      "S Xu, Z Dai, P Guo, X Fu, S Liu, L Zhou, W Tang, T Feng, M Chen, L\n",
      "Zhan, T Wu, E Hu, Y Jiang, X Bo, G Yu. ggtreeExtra: Compact\n",
      "visualization of richly annotated phylogenetic data. Molecular Biology\n",
      "and Evolution. 2021, 38(9):4039-4042. doi: 10.1093/molbev/msab166 \u001b[39m\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(treeio)\n",
    "library(ggtree)\n",
    "library(ggplot2)\n",
    "library(ape)\n",
    "library(ComplexHeatmap)\n",
    "library(cowplot)\n",
    "library(circlize)\n",
    "library(ggtreeExtra)\n",
    "library(ggnewscale)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "dcf9646f-84c9-4a10-a23a-357dae032ef6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mScale for \u001b[32my\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32my\u001b[39m, which will replace the existing scale.\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mThe column of \u001b[32mx\u001b[39m aesthetic only have one unique value with `geom = geom_tile`,\n",
      "and the `width` of `geom_tile()` is not provided, the `pwidth` will be as\n",
      "`width`.”\n"
     ]
    }
   ],
   "source": [
    "path = '/home/wangkun/mtDNA_simulation/data/'\n",
    "\n",
    "lines <- readLines(paste0(path, \"tree_color.txt\"))\n",
    "# Extract the tip_label and category\n",
    "data <- do.call(rbind, lapply(lines, function(line) {\n",
    "  parts <- strsplit(line, \" \")[[1]]\n",
    "  tip_label <- parts[1]\n",
    "  category <- parts[2]\n",
    "  return(c(tip_label, category))\n",
    "}))\n",
    "# Convert to data.frame\n",
    "df <- data.frame(tip_label = data[, 1], category = data[, 2], stringsAsFactors = FALSE)\n",
    "\n",
    "file = 'embryo_mtSNV_final_nofilter.phy_nj.nwk'\n",
    "tree <- read.tree(paste0(path, file))\n",
    "tree_plot <- ggtree(tree,layout = 'fan', open.angle = 0,branch.length='none',size=0.5)\n",
    "tree_plot$data[tree_plot$data$isTip == TRUE, 'cluster'] <- df$category[match(tree_plot$data$label[tree_plot$data$isTip == TRUE], df$tip_label)]\n",
    "\n",
    "categories <- unique(df$category)\n",
    "colors <- c('#5BD17E','#D6D0FA','#FA937F','#C6E7CE','#E48DD0','#FFC33C','#8AE0DF','#82B329','#7475E6','#FBE2B4','#8F4851','#2040BD','#A5BFFA','#B500EB','#00992B','#724094','#FC6621','#CC2A27','#C2459F','#670755')\n",
    "color_vector <- setNames(colors[1:length(categories)], categories)\n",
    "\n",
    "\n",
    "options(repr.plot.width=6, repr.plot.height=6)\n",
    "plot <- tree_plot +\n",
    "     new_scale_fill() +\n",
    "     geom_fruit(\n",
    "         geom=geom_tile,\n",
    "         mapping=aes(fill=cluster),\n",
    "         pwidth=max(tree_plot$data$x)*0.25,\n",
    "         offset=0.15\n",
    "     ) + scale_fill_manual(\n",
    "         name=\"cluster\",\n",
    "         breaks=categories,\n",
    "         values=color_vector\n",
    "     )\n",
    "\n",
    "ggsave(paste0(path, 'tree_nofilter.pdf'), plot = plot, width = 6, height = 6, device = cairo_pdf, units = \"in\", limitsize = FALSE)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad839f9d-320d-4cbd-8021-a0e7adb74f43",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ecb2189f-c0b4-48bb-8e15-5d598ed447a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mScale for \u001b[32my\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32my\u001b[39m, which will replace the existing scale.\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mThe column of \u001b[32mx\u001b[39m aesthetic only have one unique value with `geom = geom_tile`,\n",
      "and the `width` of `geom_tile()` is not provided, the `pwidth` will be as\n",
      "`width`.”\n"
     ]
    }
   ],
   "source": [
    "path = '/home/wangkun/mtDNA_simulation/data/'\n",
    "\n",
    "lines <- readLines(paste0(path, \"THQ_tree_color.txt\"))\n",
    "# Extract the tip_label and category\n",
    "data <- do.call(rbind, lapply(lines, function(line) {\n",
    "  parts <- strsplit(line, \" \")[[1]]\n",
    "  tip_label <- parts[1]\n",
    "  category <- parts[2]\n",
    "  return(c(tip_label, category))\n",
    "}))\n",
    "# Convert to data.frame\n",
    "df <- data.frame(tip_label = data[, 1], category = data[, 2], stringsAsFactors = FALSE)\n",
    "\n",
    "file = 'THQ_mtSNV_final.phy_nj.nwk'\n",
    "tree <- read.tree(paste0(path, file))\n",
    "tree_plot <- ggtree(tree,layout = 'fan', open.angle = 0,branch.length='none',size=0.5)\n",
    "tree_plot$data[tree_plot$data$isTip == TRUE, 'cluster'] <- df$category[match(tree_plot$data$label[tree_plot$data$isTip == TRUE], df$tip_label)]\n",
    "\n",
    "categories <- unique(df$category)\n",
    "colors <- c('#5BD17E','#D6D0FA','#FA937F','#C6E7CE','#E48DD0','#FFC33C','#8AE0DF','#82B329','#7475E6','#FBE2B4','#8F4851','#2040BD','#A5BFFA','#B500EB','#00992B','#724094','#FC6621','#CC2A27','#C2459F','#670755')\n",
    "color_vector <- setNames(colors[1:length(categories)], categories)\n",
    "\n",
    "\n",
    "options(repr.plot.width=6, repr.plot.height=6)\n",
    "plot <- tree_plot +\n",
    "     new_scale_fill() +\n",
    "     geom_fruit(\n",
    "         geom=geom_tile,\n",
    "         mapping=aes(fill=cluster),\n",
    "         pwidth=max(tree_plot$data$x)*0.25,\n",
    "         offset=0.15\n",
    "     ) + scale_fill_manual(\n",
    "         name=\"cluster\",\n",
    "         breaks=categories,\n",
    "         values=color_vector\n",
    "     )\n",
    "\n",
    "ggsave(paste0(path, 'THQ_mtSNV_final.pdf'), plot = plot, width = 6, height = 6, device = cairo_pdf, units = \"in\", limitsize = FALSE)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b44a0547-fed7-4aa0-bac8-3ce2aec6ab91",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d969aed8-fc75-41cf-be2a-d0a9d609a695",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41d91dcc-ef1b-4cb7-b2e9-1c80c1bff33e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.1"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": false,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
